File: 백준/Gold/14502. 연구소/연구소.cc
#include<iostream>
#include<queue>

using namespace std;

int N, M;
int Map[8][8];
int copyMap[8][8];
int dx[4] = {0,0,1,-1};
int dy[4] = {1,-1,0,0};
int ans = 0;

void bfs(){
    int cnt = 0;
    queue<pair<int,int>> q;

    for(int i=0;i<N;i++){
        for(int j=0;j<M;j++){
            if(copyMap[i][j]==2) q.push(make_pair(i,j));
        }
    }

    while(!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for(int i=0;i<4;i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(0 <= nx && nx < N && 0 <= ny && ny < M ){
                if(copyMap[nx][ny] == 0){
                    copyMap[nx][ny] = 2;
                    q.push(make_pair(nx,ny));
                }
            }
        }
    }

    for(int i=0;i<N;i++){
        for(int j=0;j<M;j++){
            if(copyMap[i][j]==0) cnt++;
        }
    }

    ans = max(ans,cnt);
}


void MakeWall(int cnt){
    if(cnt ==3){
        copy(&Map[0][0], &Map[0][0] + 64, &copyMap[0][0]);
        bfs();
        return;
    }
    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            if(Map[i][j] == 0){
                Map[i][j] = 1;
                MakeWall(cnt + 1);
                Map[i][j] = 0;
            }
        }
    }
}

int main(){

    scanf("%d %d", &N, &M);
    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            scanf("%d", &Map[i][j]);
        }
    }

    MakeWall(0);
    printf("%d\n", ans);



    return 0;
}

File: 백준/Gold/14502. 연구소/연구소.py
#Gold IV
from collections import deque
import copy
N, M = map(int, input().split())
Origin = [list(map(int, input().split())) for _ in range(N)]
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
ans = 0

def bfs():
    global ans
    q = deque()
    Map = copy.deepcopy(Origin)

    for i in range(N):
        for j in range(M):
            if Map[i][j] == 2:
                q.append([i,j])

    while len(q) > 0:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < M:
                if Map[nx][ny] == 0:
                    Map[nx][ny] = 2
                    q.append([nx, ny])

    cnt = 0
    for i in range(N):
        cnt += Map[i].count(0)

    ans = max(ans,cnt)

def MakeWall(cnt):
    if cnt == 3:
        bfs()
        return

    for i in range(N):
        for j in range(M):
            if Origin[i][j] == 0:
                Origin[i][j] = 1
                MakeWall(cnt+1)
                Origin[i][j] = 0

MakeWall(0)
print(ans)